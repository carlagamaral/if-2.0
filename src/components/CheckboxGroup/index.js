import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Checkbox from '../Checkbox';
import CheckboxGroupWrapper from './styles';

const CheckboxGroup = ({ itemIds, onClick, startChecked }) => {
	const [checkboxes, setCheckboxes] = useState([]);

	useEffect(() => {
		setCheckboxes(itemIds.map((id) => ({ id, isChecked: startChecked })));
	}, [itemIds, startChecked]);

	const handleClick = (checkboxId) => {
		onClick(checkboxId);
		const updatedCheckboxes = checkboxes.map((checkbox) =>
			checkbox.id === checkboxId
				? { ...checkbox, isChecked: !checkbox.isChecked }
				: checkbox
		);

		setCheckboxes(updatedCheckboxes);
	};

	return (
		<CheckboxGroupWrapper data-testid="checkbox-group">
			{checkboxes.map((checkbox) => (
				<Checkbox
					click={() => handleClick(checkbox.id)}
					isChecked={checkbox.isChecked}
					key={checkbox.id}
					label={checkbox.id}
					testId={checkbox.id}
				/>
			))}
		</CheckboxGroupWrapper>
	);
};

CheckboxGroup.propTypes = {
	itemIds: PropTypes.arrayOf(PropTypes.string).isRequired,
	onClick: PropTypes.func,
	startChecked: PropTypes.bool,
};

CheckboxGroup.defaultProps = {
	onClick: () => {},
	startChecked: false,
};

export default CheckboxGroup;
