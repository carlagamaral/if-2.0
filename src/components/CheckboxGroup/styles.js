import styled from 'styled-components';

export const CheckboxGroupWrapper = styled.div`
	display: grid;
	grid-template-columns: 1fr 1fr;
	grid-auto-rows: 50px;
	column-gap: 4em;
`;
