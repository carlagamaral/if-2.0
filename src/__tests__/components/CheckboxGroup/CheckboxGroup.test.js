/* eslint-disable function-paren-newline */
import 'jest-dom/extend-expect'
import { cleanup, waitForElement, render, fireEvent } from 'react-testing-library'
import React from 'react'
import CheckboxGroup from '../../../components/CheckboxGroup'

describe( 'CheckboxGroup tests ', () => {
  const items = [ 'A', 'B', 'C' ]

  afterEach( () => {
    cleanup()
    jest.clearAllMocks()
  } )

  test( 'Renders one checkbox for each item', async() => {
    const { container, getByTestId } = render(
      <CheckboxGroup itemIds={ items } />
    )

    const checkboxGroup = await waitForElement( () => getByTestId( 'checkbox-group' ) )

    for ( const item of items ) {
      // eslint-disable-next-line no-await-in-loop
      const checkbox = await waitForElement( () => getByTestId( item ) )

      expect( checkbox ).toBeInTheDocument()
    }

    expect( container ).toMatchSnapshot()

    expect( checkboxGroup ).toBeInTheDocument()
    expect( checkboxGroup.childElementCount ).toEqual( items.length )
  } )

  test( 'It can check and uncheck a children checkbox', async() => {
    const { getByTestId } = render(
      <CheckboxGroup itemIds={ items } />
    )

    const checkboxA = await waitForElement( () => getByTestId( 'A' ) )

    fireEvent.click( checkboxA )
    expect( checkboxA.checked ).toBeTruthy()

    fireEvent.click( checkboxA )
    expect( checkboxA.checked ).toBeFalsy()
  } )

  test( 'It can check more than one children checkbox', async() => {
    const { getByTestId } = render(
      <CheckboxGroup itemIds={ items } />
    )

    const checkboxA = await waitForElement( () => getByTestId( 'A' ) )
    const checkboxC = await waitForElement( () => getByTestId( 'C' ) )

    fireEvent.click( checkboxA )
    fireEvent.click( checkboxC )

    expect( checkboxA.checked ).toBeTruthy()
    expect( checkboxC.checked ).toBeTruthy()
  } )

  test( 'When clicking a checkbox it calls the onClick function', async() => {
    const mockedFunction = jest.fn()

    const { getByTestId } = render(
      <CheckboxGroup
        itemIds={ items }
        onClick={ mockedFunction }
      />
    )

    const checkboxA = await waitForElement( () => getByTestId( 'A' ) )

    fireEvent.click( checkboxA )

    expect( mockedFunction ).toHaveBeenCalledWith( 'A' )
  } )
} )
