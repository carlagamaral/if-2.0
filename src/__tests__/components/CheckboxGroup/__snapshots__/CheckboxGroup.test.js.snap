// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup tests  Renders one checkbox for each item 1`] = `
<div>
  <div
    class="sc-AxirZ laTrJs"
    data-testid="checkbox-group"
  >
    <fieldset
      class="sc-AxjAm iQWkcR"
    >
      <div
        class="checkbox "
        data-testid="div-A"
      >
        <label>
          <input
            data-testid="A"
            type="checkbox"
            value="A"
          />
          <span
            class="checkbox-material"
          >
            <span
              class="check"
            />
          </span>
          A
        </label>
      </div>
    </fieldset>
    <fieldset
      class="sc-AxjAm iQWkcR"
    >
      <div
        class="checkbox "
        data-testid="div-B"
      >
        <label>
          <input
            data-testid="B"
            type="checkbox"
            value="B"
          />
          <span
            class="checkbox-material"
          >
            <span
              class="check"
            />
          </span>
          B
        </label>
      </div>
    </fieldset>
    <fieldset
      class="sc-AxjAm iQWkcR"
    >
      <div
        class="checkbox "
        data-testid="div-C"
      >
        <label>
          <input
            data-testid="C"
            type="checkbox"
            value="C"
          />
          <span
            class="checkbox-material"
          >
            <span
              class="check"
            />
          </span>
          C
        </label>
      </div>
    </fieldset>
  </div>
</div>
`;
